lib/SILOptimizer/Analysis/AccessSummaryAnalysis.cpp

```diff
@@ -0,0 +1,329 @@
+//===--- AccessSummaryAnalysis.cpp - SIL Access Summary Analysis ----------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "sil-access-summary-analysis"
+#include "swift/SIL/SILArgument.h"
+#include "swift/SILOptimizer/Analysis/AccessSummaryAnalysis.h"
+#include "swift/SILOptimizer/Analysis/FunctionOrder.h"
+#include "swift/SILOptimizer/PassManager/PassManager.h"
+#include "llvm/ADT/StringExtras.h"
+
+using namespace swift;
+
+void AccessSummaryAnalysis::processFunction(FunctionInfo *info,
+                                            FunctionOrder &order) {
+  // Does the summary need to be recomputed?
+  if (order.prepareForVisiting(info))
+    return;
+
+  // Compute function summary on a per-parameter basis.
+  unsigned index = 0;
+  for (SILArgument *arg : info->getFunction()->getArguments()) {
+    FunctionSummary &functionSummary = info->getSummary();
+    ParameterSummary &paramSummary =
+        functionSummary.getAccessForParameter(index);
+    index++;
+
+    auto *functionArg = cast<SILFunctionArgument>(arg);
+    // Only summarize @inout_aliasable parameters.
+    SILArgumentConvention convention =
+        functionArg->getArgumentConvention().Value;
+    if (convention != SILArgumentConvention::Indirect_InoutAliasable)
+      continue;
+
+    processParameter(info, functionArg, paramSummary, order);
+  }
+}
+
+/// Track uses of the parameter, recording in the summary any accesses
+/// started by a begin_access and any flows of the parameter to other
+/// functions.
+void AccessSummaryAnalysis::processParameter(FunctionInfo *info,
+                                             SILFunctionArgument *parameter,
+                                             ParameterSummary &summary,
+                                             FunctionOrder &order) {
+  unsigned parameterIndex = parameter->getIndex();
```

Comment:
What you're calling "parameter" is actually a SILFunctionArgument and
when you get its index you're getting the function argument index
(including indirect results), not the parameter index (user to index
into the function signature via SILFunctionType). This is a particular
pain point for me because I've spent several weeks worth of time
auditing and debugging code that doesn't make a proper distinction
between argument and parameter indices. For a given function
signature, argument indices change as SIL is lowered while parameter
indices do not.

I realize you only care about parameters here, so if you really want
to talk about parameters, then you can need to adjust the index
accordingly using
getFunction()->getConventions()->getSILArgIndexOfFirstParam(). But for
you, it's easier to just work with the SILArgument index and call it
what it is. If you were basing this on other code in SILOpt, point me
to it I'll cleanup that code too.

You're already doing the right thing to convert "applied argument
indices" to "function argument indices", although we lack a nice
interface for that.